File: ports/kotlin/CyclePredictor.kt
// Cycle predictor — Kotlin port of the reference kernel (src/cycle_predictor/portable.py).
// Pure Kotlin stdlib (+ java.time for the date wrapper); suits Android and JVM backends.
// Validated against artifacts/test_vectors.json in CI (see ports/kotlin/Validate.kt).
import java.time.LocalDate
import java.time.temporal.ChronoUnit
import kotlin.math.PI
import kotlin.math.exp
import kotlin.math.ln
import kotlin.math.pow
import kotlin.math.roundToInt
import kotlin.math.sqrt

data class Backbone(
    val muLog: Double, val tau: Double, val pi: Double, val xi: Double,
    val sMax: Int, val lamMin: Double, val lamMax: Double, val grid: Int,
)
data class TwoPhase(val lutealMean: Double, val lutealSd: Double)
data class ThermalShift(val baseline: Int, val run: Int, val delta: Double, val search: List<Int>)
data class Params(
    val version: String, val backbone: Backbone, val twophase: TwoPhase,
    val lhSurge: Double, val thermalShift: ThermalShift,
)
data class KernelResult(val cycleLength: Double, val sd: Double, val mode: String)

private fun logsumexp(vals: List<Double>): Double {
    var m = Double.NEGATIVE_INFINITY
    for (v in vals) if (v > m) m = v
    if (m == Double.NEGATIVE_INFINITY) return m
    var s = 0.0
    for (v in vals) s += exp(v - m)
    return m + ln(s)
}

/** Generalized-Poisson posterior-predictive (mean, sd) for the next cycle length. */
fun backbonePredict(bp: Backbone, history: List<Double>): Pair<Double, Double> {
    val g = bp.grid
    val lo = ln(bp.lamMin); val hi = ln(bp.lamMax)
    val loglam = DoubleArray(g); val lam = DoubleArray(g)
    for (i in 0 until g) { val x = lo + (hi - lo) * i / (g - 1); loglam[i] = x; lam[i] = exp(x) }

    val xi = bp.xi; val pi = bp.pi
    val s = (0..bp.sMax).toList()
    val logwUn = s.map { ln(1 - pi) + it * ln(pi) }
    val zz = logsumexp(logwUn)
    val logw = logwUn.map { it - zz }
    val mult = s.map { (1 + it).toDouble() }

    val logpost = DoubleArray(g) { -0.5 * ((loglam[it] - bp.muLog) / bp.tau).pow(2) - ln(bp.tau) - 0.5 * ln(2 * PI) }
    for (d in history) {
        val di = maxOf(1, d.roundToInt()).toDouble()
        for (gi in 0 until g) {
            val terms = ArrayList<Double>(s.size)
            for (si in s.indices) {
                val a = mult[si] * lam[gi]
                val arg = a + di * xi
                terms.add(if (arg > 0) logw[si] + ln(a) + (di - 1) * ln(arg) - arg else Double.NEGATIVE_INFINITY)
            }
            logpost[gi] += logsumexp(terms)
        }
    }
    val z2 = logsumexp(logpost.toList())
    var elam = 0.0; var elam2 = 0.0
    for (i in 0 until g) { val p = exp(logpost[i] - z2); elam += p * lam[i]; elam2 += p * lam[i] * lam[i] }
    val varLam = maxOf(elam2 - elam * elam, 0.0)
    val c1 = 1.0 / (1.0 - xi); val c3 = 1.0 / (1.0 - xi).pow(3)
    return Pair(elam * c1, sqrt(elam * c3 + varLam * c1 * c1))
}

fun detectThermalShift(temp: Map<Int, Double>, ts: ThermalShift): Int? {
    val baseline = ts.baseline; val run = ts.run
    val lo = ts.search[0]; val hi = ts.search[1]
    if (temp.size < baseline + run) return null
    for (d in lo..hi) {
        val base = ArrayList<Double>(); for (k in (d - baseline) until d) temp[k]?.let { base.add(it) }
        val fut = ArrayList<Double>(); for (j in 0 until run) temp[d + j]?.let { fut.add(it) }
        if (base.size < baseline - 1 || fut.size < run) continue
        val cover = base.sum() / base.size + ts.delta
        if (fut.all { it > cover }) return maxOf(d - 1, 0)
    }
    return null
}

/** Numpy-free equivalent of UnifiedPredictor.predict. Signals keyed by day-of-cycle. */
fun predict(
    params: Params, history: List<Double>,
    lhByDay: Map<Int, Double>? = null, tempByDay: Map<Int, Double>? = null,
): KernelResult {
    if (!lhByDay.isNullOrEmpty()) {
        val surge = lhByDay.filter { it.value >= params.lhSurge }.keys.sorted()
        if (surge.isNotEmpty())
            return KernelResult(surge[0] + params.twophase.lutealMean, params.twophase.lutealSd, "two_phase_lh")
    }
    if (!tempByDay.isNullOrEmpty()) {
        val est = detectThermalShift(tempByDay, params.thermalShift)
        if (est != null)
            return KernelResult(est + params.twophase.lutealMean, params.twophase.lutealSd, "two_phase_wearable")
    }
    val (mean, sd) = backbonePredict(params.backbone, history)
    return KernelResult(mean, sd, "history")
}

// --- date wrapper (app-facing; mirrors cycle_predictor/api.py, not in the kernel vectors) ---
data class UserLog(
    val periodStarts: List<String>,
    val lhTests: Map<String, Double>? = null,
    val wearableTemp: Map<String, Double>? = null,
    val today: String? = null,
)
data class PeriodForecast(
    val predictedStart: LocalDate, val earliest: LocalDate, val latest: LocalDate,
    val confidence: Double, val daysUntil: Int, val nHistory: Int,
    val cycleLengthDays: Double, val sdDays: Double, val mode: String,
)

// Inverse standard-normal CDF (Acklam) — for the confidence interval only.
fun invNormCdf(p: Double): Double {
    val a = doubleArrayOf(-3.969683028665376e1, 2.209460984245205e2, -2.759285104469687e2, 1.383577518672690e2, -3.066479806614716e1, 2.506628277459239)
    val b = doubleArrayOf(-5.447609879822406e1, 1.615858368580409e2, -1.556989798598866e2, 6.680131188771972e1, -1.328068155288572e1)
    val c = doubleArrayOf(-7.784894002430293e-3, -3.223964580411365e-1, -2.400758277161838, -2.549732539343734, 4.374664141464968, 2.938163982698783)
    val d = doubleArrayOf(7.784695709041462e-3, 3.224671290700398e-1, 2.445134137142996, 3.754408661907416)
    val pl = 0.02425
    return when {
        p < pl -> { val q = sqrt(-2 * ln(p)); (((((c[0]*q+c[1])*q+c[2])*q+c[3])*q+c[4])*q+c[5]) / ((((d[0]*q+d[1])*q+d[2])*q+d[3])*q+1) }
        p <= 1 - pl -> { val q = p - 0.5; val r = q * q; (((((a[0]*r+a[1])*r+a[2])*r+a[3])*r+a[4])*r+a[5]) * q / (((((b[0]*r+b[1])*r+b[2])*r+b[3])*r+b[4])*r+1) }
        else -> { val q = sqrt(-2 * ln(1 - p)); -(((((c[0]*q+c[1])*q+c[2])*q+c[3])*q+c[4])*q+c[5]) / ((((d[0]*q+d[1])*q+d[2])*q+d[3])*q+1) }
    }
}

fun predictNextPeriod(log: UserLog, params: Params, confidence: Double = 0.8): PeriodForecast {
    val starts = log.periodStarts.map { LocalDate.parse(it) }.sorted()
    require(starts.isNotEmpty()) { "need at least one period start date" }
    val onset = starts.last()
    val history = (1 until starts.size).map { ChronoUnit.DAYS.between(starts[it - 1], starts[it]).toDouble() }
    val today = log.today?.let { LocalDate.parse(it) } ?: onset

    fun offsets(m: Map<String, Double>?): Map<Int, Double>? {
        if (m == null) return null
        val out = HashMap<Int, Double>()
        for ((k, v) in m) { val off = ChronoUnit.DAYS.between(onset, LocalDate.parse(k)).toInt(); if (off >= 0) out[off] = v }
        return out
    }
    val r = predict(params, history, offsets(log.lhTests), offsets(log.wearableTemp))
    val half = invNormCdf(0.5 + confidence / 2) * r.sd
    val predicted = onset.plusDays(r.cycleLength.roundToInt().toLong())
    return PeriodForecast(
        predictedStart = predicted,
        earliest = onset.plusDays((r.cycleLength - half).roundToInt().toLong()),
        latest = onset.plusDays((r.cycleLength + half).roundToInt().toLong()),
        confidence = confidence,
        daysUntil = ChronoUnit.DAYS.between(today, predicted).toInt(),
        nHistory = history.size,
        cycleLengthDays = (r.cycleLength * 10).roundToInt() / 10.0,
        sdDays = (r.sd * 10).roundToInt() / 10.0,
        mode = r.mode,
    )
}

File: ports/typescript/cyclePredictor.ts
// Cycle predictor — TypeScript port of the reference kernel (src/cycle_predictor/portable.py).
// Dependency-free; runs in browsers and Node. Validated against artifacts/test_vectors.json
// (see validate.ts). Load params from artifacts/model.json.

export interface BackboneParams {
  type?: string;
  mu_log: number; tau: number; pi: number; xi: number;
  s_max: number; lam_min: number; lam_max: number; grid: number;
}
export interface Params {
  version: string;
  backbone: BackboneParams;
  twophase: { luteal_mean: number; luteal_sd: number };
  lh_surge: number;
  thermal_shift: { baseline: number; run: number; delta: number; search: [number, number] };
}
export type Mode = "history" | "two_phase_lh" | "two_phase_wearable";
export interface KernelResult { cycleLength: number; sd: number; mode: Mode; }

function logsumexp(vals: number[]): number {
  let m = -Infinity;
  for (const v of vals) if (v > m) m = v;
  if (m === -Infinity) return -Infinity;
  let s = 0;
  for (const v of vals) s += Math.exp(v - m);
  return m + Math.log(s);
}

/** Generalized-Poisson posterior-predictive [mean, sd] for the next cycle length. */
export function backbonePredict(bp: BackboneParams, history: number[]): [number, number] {
  const G = bp.grid;
  const lo = Math.log(bp.lam_min), hi = Math.log(bp.lam_max);
  const loglam: number[] = [], lam: number[] = [];
  for (let i = 0; i < G; i++) { const x = lo + (hi - lo) * i / (G - 1); loglam.push(x); lam.push(Math.exp(x)); }

  const { xi, pi, s_max: sMax, mu_log: muLog, tau } = bp;
  const S: number[] = []; for (let s = 0; s <= sMax; s++) S.push(s);
  const logwUn = S.map((s) => Math.log(1 - pi) + s * Math.log(pi));
  const zz = logsumexp(logwUn);
  const logw = logwUn.map((w) => w - zz);
  const mult = S.map((s) => 1 + s);

  const logpost = loglam.map((x) => -0.5 * ((x - muLog) / tau) ** 2 - Math.log(tau) - 0.5 * Math.log(2 * Math.PI));
  for (const d of history) {
    const di = Math.max(1, Math.round(d));
    for (let gi = 0; gi < G; gi++) {
      const terms: number[] = [];
      for (let si = 0; si < S.length; si++) {
        const a = mult[si] * lam[gi];
        const arg = a + di * xi;
        terms.push(arg > 0 ? logw[si] + Math.log(a) + (di - 1) * Math.log(arg) - arg : -Infinity);
      }
      logpost[gi] += logsumexp(terms);
    }
  }
  const z2 = logsumexp(logpost);
  let elam = 0, elam2 = 0;
  for (let i = 0; i < G; i++) { const p = Math.exp(logpost[i] - z2); elam += p * lam[i]; elam2 += p * lam[i] * lam[i]; }
  const varLam = Math.max(elam2 - elam * elam, 0);
  const c1 = 1 / (1 - xi), c3 = 1 / (1 - xi) ** 3;
  return [elam * c1, Math.sqrt(elam * c3 + varLam * c1 * c1)];
}

function toDayMap(obj: Record<string | number, number> | null | undefined): Map<number, number> {
  const m = new Map<number, number>();
  if (obj) for (const [k, v] of Object.entries(obj)) if (v != null) m.set(Number(k), v);
  return m;
}

export function detectThermalShift(temp: Map<number, number>, ts: Params["thermal_shift"]): number | null {
  const { baseline, run, delta } = ts; const [lo, hi] = ts.search;
  if (temp.size < baseline + run) return null;
  for (let d = lo; d <= hi; d++) {
    const base: number[] = []; for (let k = d - baseline; k < d; k++) if (temp.has(k)) base.push(temp.get(k)!);
    const fut: number[] = []; for (let j = 0; j < run; j++) if (temp.has(d + j)) fut.push(temp.get(d + j)!);
    if (base.length < baseline - 1 || fut.length < run) continue;
    const cover = base.reduce((a, b) => a + b, 0) / base.length + delta;
    if (fut.every((t) => t > cover)) return Math.max(d - 1, 0);
  }
  return null;
}

/** Numpy-free equivalent of UnifiedPredictor.predict. Signals keyed by day-of-cycle. */
export function predict(
  params: Params, history: number[],
  lhByDay?: Record<string | number, number> | null,
  tempByDay?: Record<string | number, number> | null,
): KernelResult {
  const lh = toDayMap(lhByDay);
  if (lh.size) {
    const surge = [...lh.entries()].filter(([, v]) => v >= params.lh_surge).map(([d]) => d).sort((a, b) => a - b);
    if (surge.length) return { cycleLength: surge[0] + params.twophase.luteal_mean, sd: params.twophase.luteal_sd, mode: "two_phase_lh" };
  }
  const temp = toDayMap(tempByDay);
  if (temp.size) {
    const est = detectThermalShift(temp, params.thermal_shift);
    if (est !== null) return { cycleLength: est + params.twophase.luteal_mean, sd: params.twophase.luteal_sd, mode: "two_phase_wearable" };
  }
  const [mean, sd] = backbonePredict(params.backbone, history);
  return { cycleLength: mean, sd, mode: "history" };
}

// --- date wrapper (mirrors cycle_predictor/api.py; not covered by the kernel vectors) ---
export interface UserLog {
  periodStarts: (string | Date)[];
  lhTests?: Record<string, number>;
  wearableTemp?: Record<string, number>;
  today?: string | Date;
}
export interface PeriodForecast {
  predictedStart: Date; earliest: Date; latest: Date; confidence: number;
  daysUntil: number; cycleLengthDays: number; sdDays: number; mode: Mode; nHistory: number;
}

const DAY = 86400000;
const asDate = (d: string | Date): Date => (d instanceof Date ? d : new Date(d + "T00:00:00Z"));
const dayDiff = (a: Date, b: Date) => Math.round((a.getTime() - b.getTime()) / DAY);
const addDays = (d: Date, n: number) => new Date(d.getTime() + n * DAY);

// Inverse standard-normal CDF (Acklam) — for the confidence interval only.
function invNormCdf(p: number): number {
  const a = [-3.969683028665376e1, 2.209460984245205e2, -2.759285104469687e2, 1.383577518672690e2, -3.066479806614716e1, 2.506628277459239];
  const b = [-5.447609879822406e1, 1.615858368580409e2, -1.556989798598866e2, 6.680131188771972e1, -1.328068155288572e1];
  const c = [-7.784894002430293e-3, -3.223964580411365e-1, -2.400758277161838, -2.549732539343734, 4.374664141464968, 2.938163982698783];
  const d = [7.784695709041462e-3, 3.224671290700398e-1, 2.445134137142996, 3.754408661907416];
  const pl = 0.02425;
  if (p < pl) { const q = Math.sqrt(-2 * Math.log(p)); return (((((c[0] * q + c[1]) * q + c[2]) * q + c[3]) * q + c[4]) * q + c[5]) / ((((d[0] * q + d[1]) * q + d[2]) * q + d[3]) * q + 1); }
  if (p <= 1 - pl) { const q = p - 0.5, r = q * q; return (((((a[0] * r + a[1]) * r + a[2]) * r + a[3]) * r + a[4]) * r + a[5]) * q / (((((b[0] * r + b[1]) * r + b[2]) * r + b[3]) * r + b[4]) * r + 1); }
  const q = Math.sqrt(-2 * Math.log(1 - p)); return -(((((c[0] * q + c[1]) * q + c[2]) * q + c[3]) * q + c[4]) * q + c[5]) / ((((d[0] * q + d[1]) * q + d[2]) * q + d[3]) * q + 1);
}

export function predictNextPeriod(log: UserLog, params: Params, confidence = 0.8): PeriodForecast {
  const starts = log.periodStarts.map(asDate).sort((a, b) => a.getTime() - b.getTime());
  if (!starts.length) throw new Error("need at least one period start date");
  const onset = starts[starts.length - 1];
  const history: number[] = [];
  for (let i = 1; i < starts.length; i++) history.push(dayDiff(starts[i], starts[i - 1]));
  const today = log.today ? asDate(log.today) : onset;

  const offsets = (m?: Record<string, number>) => {
    if (!m) return null;
    const out: Record<number, number> = {};
    for (const [d, v] of Object.entries(m)) { const off = dayDiff(asDate(d), onset); if (off >= 0) out[off] = v; }
    return out;
  };
  const r = predict(params, history, offsets(log.lhTests), offsets(log.wearableTemp));
  const z = invNormCdf(0.5 + confidence / 2);
  const half = z * r.sd;
  return {
    predictedStart: addDays(onset, Math.round(r.cycleLength)),
    earliest: addDays(onset, Math.round(r.cycleLength - half)),
    latest: addDays(onset, Math.round(r.cycleLength + half)),
    confidence,
    daysUntil: dayDiff(addDays(onset, Math.round(r.cycleLength)), today),
    cycleLengthDays: Math.round(r.cycleLength * 10) / 10,
    sdDays: Math.round(r.sd * 10) / 10,
    mode: r.mode,
    nHistory: history.length,
  };
}
